vm/isolated/isolated.go

```diff
@@ -173,26 +182,38 @@ func (inst *instance) ssh(command string) error {
 
 func (inst *instance) repair() error {
 	log.Logf(2, "isolated: trying to ssh")
-	if err := inst.waitForSSH(30 * time.Minute); err == nil {
+	if err := inst.waitForSSH(3 * time.Minute); err == nil {
+		log.Logf(2, "isolated: ssh succeeded")
+	} else {
+		log.Logf(2, "isolated: ssh failed")
 		if inst.cfg.TargetReboot {
 			log.Logf(2, "isolated: trying to reboot")
-			inst.ssh("reboot") // reboot will return an error, ignore it
-			if err := inst.waitForReboot(5 * 60); err != nil {
+			if inst.cfg.USBDevNum == "" {
+				inst.ssh("reboot")
+			} else {
+				_, err = exec.Command("/bin/sh", "-c", "echo 0 > /sys/bus/usb/devices/"+inst.cfg.USBDevNum+"/authorized").Output()
+				if err != nil {
+					log.Logf(2, "isolated: failed to turn off the device")
+					return err
+				}
+				_, err = exec.Command("/bin/sh", "-c", "echo 1 > /sys/bus/usb/devices/"+inst.cfg.USBDevNum+"/authorized").Output()
+				if err != nil {
+					log.Logf(2, "isolated: failed to turn on the device")
+					return err
+				}
+			}
+
+			if err := inst.waitForReboot(3 * 60); err != nil {
 				log.Logf(2, "isolated: machine did not reboot")
 				return err
 			}
 			log.Logf(2, "isolated: rebooted wait for comeback")
-			if err := inst.waitForSSH(30 * time.Minute); err != nil {
-				log.Logf(2, "isolated: machine did not comeback")
+			if err := inst.waitForSSH(3 * time.Minute); err != nil {
```

Comment:
Previously we waited for up to 30 minutes. Now we wait significantly less. This is used for machines that reboot for 10+ minutes. Please keep the old timeouts at least.
You may need to restructure logic somewhat. If we can power cycle it, we don't need to wait for ssh in the beginning at all. First power cycle and then wait for ssh. In this case it won't harm to wait for a long time.

Reply:
The reason why I changed the waiting time is,
when `syz-manager` is started, it calls `Create()` in `isolated.go`.
and then `Create()` calls `repair()`, `repair()` do reboot the machine.
Then, it should wait 30 minutes always even though it is rebooted.
Because in `waitForSSH()`  function, there is the code as followings in while loop:
		if err = inst.ssh("pwd"); err != nil {
			return nil
		}
these lines do ssh command `pwd ` to check whether the target is rebooted. (maybe?)
if the machine is rebooted,  the `err` is `nil`.
If `err` is `nil`, it should return `nil` and get out the while loop.
but the code return `nil` when `err` is not `nil`.
so, it cannot get out the while loop even though the target is rebooted.

so, when I start the fuzzing, I always should wait 30 minutes.
So, I changed the code.

Now, I ask that 
how about the change the code?
I'm not sure that I'm right.

I think the code should be like this:
		if err = inst.ssh("pwd"); err == nil {
			return nil
		}

Reply:
I don't see this code in vmimpl.WaitForSSH. I see this code in waitForReboot, but waitForReboot is not executed with 30min timeout. Also that line has a comment that explains its purpose:
```
// If it fails, then the reboot started
```
If you do the reboot, ssh should fail and waitForReboot should return immediately.
What am I missing?

Reply:
Sorry, `waitForReboot` is right.

Also, I tried to figure out this issue.
Now, I got more information, but I can't understand. 
There is my log, check it out. 

- run `syz-manager`
- `Create()` -> `repair()` -> `inst.ssh("reboot")`

in `ssh()`,
- `cmd.Start()` -> reboot the target machine
- `cmd.Wait()`

-> the rebooting target machine is completed within 25 secs.

but `ssh()` still wait on `cmd.Wait()`.
I don't know why.
it did not go to 
```
       close(done)
	if inst.debug {
		log.Logf(0, "ssh returned")
	}
	return nil
```

then after 30 secs, it executes the code in 
```
		case <-time.After(time.Second * 30):
			if inst.debug {
				log.Logf(0, "ssh hanged")
			}
			cmd.Process.Kill()
```

it kills the the process.
Now, `cmd.Wait()` is end because it is killed. 
so, it returns error `return fmt.Errorf("ssh %+v failed: %v\n%s", args, err, out)`

now, it come back to `repair()`.
and executes `inst.waitForReboot()`
in `waitForReboot()`,
it executes `inst.ssh("pwd")`,
but the target machine is already rebooted, 
so it does not return error.
so it cannot get out the for loop until timeout.

this is my log.
So, the main problem is `ssh()`.

`cmd.Wait()` is not end even though the target machine is rebooted.

I try to study about it, but I'm a GO lang newbie, I can't understand why.
I believe that you guys can figure it out.

Reply:
Humm... I see.
From what you describe it looks like the remove machine has started shutdown, ungracefully killed sshd, and now the local ssh hanged b/c it does not receive any data from remove side (so it looks like slow connection).
But I can't reproduce such behavior locally. When I try to do "ssh reboot" for my VM, the ssh command returns immediately.
Some possible solutions that I can think of:
1. Try to use ssh ServerAliveInterval/ServerAliveCountMax options (see below). Then we could set connection timeout to 5 secs or so, so that "ssh reboot" returns even if remote side is dead.
2. Do graceful reboot so that sshd properly terminates connection. However normally reboot should already do a graceful shutdown...
3. Use delayed reboot so that ssh properly finished before reboot has started. Unfortunately, shutdown command seems to only be able to specify delay in minutes. Or could be run something like `bash -c "(sleep 5; reboot )&", but the async part (&) should apply to both sleep and reboot, not sure how to express that in bash. The idea is to kick off reboot after 5 secs and return.
4. If nothing of that works, we could do something more explicit: start reboot with short timeout on the ssh command, try see if new ssh'es fail (which would mean reboot has started); if they fail we rebooted it; otherwise, try to reboot again.
5. Or, we need reboot only if we don't have the USB hub. And you have it. And I think we need slightly different code paths for USB/no-USB anyway. So you could simply keep the current code as is, but if USB is present, then use new logic (it won't need to infer if reboot has started or not). I don't know if the current code is broken or not or somebody uses it in slightly different setup, but if you don't touch it, you don't need to fix it :) We need to test all this code very much, but I can't figure out a reasonable testing story...

```
     ServerAliveInterval
             Sets a timeout interval in seconds after which if no data has been received from the
             server, ssh(1) will send a message through the encrypted channel to request a re‐
             sponse from the server.  The default is 0, indicating that these messages will not
             be sent to the server.

     ServerAliveCountMax
             Sets the number of server alive messages (see below) which may be sent without
             ssh(1) receiving any messages back from the server.  If this threshold is reached
             while server alive messages are being sent, ssh will disconnect from the server,
             terminating the session.  It is important to note that the use of server alive mes‐
             sages is very different from TCPKeepAlive (below).  The server alive messages are
             sent through the encrypted channel and therefore will not be spoofable.  The TCP
             keepalive option enabled by TCPKeepAlive is spoofable.  The server alive mechanism
             is valuable when the client or server depend on knowing when a connection has become
             inactive.

             The default value is 3.  If, for example, ServerAliveInterval (see below) is set to
             15 and ServerAliveCountMax is left at the default, if the server becomes unrespon‐
             sive, ssh will disconnect after approximately 45 seconds.
```